backend/src/service/courseService.js

```diff
@@ -501,6 +501,77 @@ const checkOutComeAfterResolve = async (req, res, next) => {
   }
 };
 
+
+// In your courseService.js file
+
+const autoAssignSubject = async (req, res, next) => {
+  try {
+    const userPlanner = req.userPlanner;
+    const planner = userPlanner.planner;
+    const subjectCode = req.params.subjectCode;
+
+    if (!subjectCode) {
+      return res.status(400).json({ error: 'No subject code provided' });
+    }
+
+    // Get the subject from the database
+    const collection = await mongoClient.getCollection('Subject');
+    const subject = await collection.findOne({ subjectCode: subjectCode });
+
+    if (!subject) {
+      return res.status(404).json({ error: 'Subject not found' });
+    }
+
+    // Get the study periods when the subject is available (e.g., ['Semester 1', 'Semester 2'])
+    const studyPeriods = subject.studyPeriod; // e.g., ['Semester 1', 'Semester 2']
+
+    if (!studyPeriods || studyPeriods.length === 0) {
+      return res.status(400).json({ error: 'No available study periods for this subject' });
+    }
+
+    // Map study periods to planner keys ('Semester 1' -> 's1', etc.)
+    const semesterMap = {
+      'Semester 1': 's1',
+      'Semester 2': 's2',
+      'Summer Term': 'su',
+      'Winter Term': 'wi'
+    };
+    
+    const availableSemesters = studyPeriods.map(period => semesterMap[period]);
+
+    // Loop over available semesters and years to find the earliest available slot
+    const years = ['y1', 'y2', 'y3'];
+    let assigned = false;
+
+    for (const sem of availableSemesters) {
+      for (const year of years) {
+        const term = `${year}${sem}`; // e.g., 'y1s1'
+        if (planner[term]) {
+          const positions = ['p1', 'p2', 'p3', 'p4'];
+          for (const pos of positions) {
+            if (!planner[term][pos] || Object.keys(planner[term][pos]).length === 0) {
+              // Assign the subject to this slot
+              planner[term][pos] = subject;
+              assigned = true;
+              break;
+            }
+          }
+        }
+        if (assigned) break;
+      }
+      if (assigned) break;
+    }
+
+    if (!assigned) {
```

Comment:
if there are no valid semester, we should assign it to an invalid semester instead of returning error. We should only return error if there are no empty slot in the planner.